ProcessLib/LIE/Common/LevelSetFunction.cpp

```diff
@@ -26,11 +34,101 @@ namespace ProcessLib
 {
 namespace LIE
 {
-double calculateLevelSetFunction(FractureProperty const& frac, double const* x_)
+
+double levelset_fracture(FractureProperty const& frac, Eigen::Vector3d const& x)
+{
+	return boost::math::sign(frac.normal_vector.dot(x - frac.point_on_fracture));
+}
+
+double levelset_branch(BranchProperty const& branch, Eigen::Vector3d const& x)
 {
-    Eigen::Map<Eigen::Vector3d const> x(x_, 3);
-    return Heaviside(
-        boost::math::sign(frac.normal_vector.dot(x - frac.point_on_fracture)));
+	return boost::math::sign(branch.normal_vector_branch.dot(x - branch.coords));
+}
+
+
+std::vector<double> u_global_enrichments(
+	std::vector<FractureProperty*> const& frac_props,
+	std::vector<JunctionProperty*> const& junction_props,
+	std::unordered_map<int,int> const& fracID_to_local,
+	Eigen::Vector3d const& x)
+{
+	//pre-calculate levelsets for all fractures
+	std::vector<double> levelsets(frac_props.size());
+	for (std::size_t i=0; i<frac_props.size(); i++)
+		levelsets[i] = Heaviside(levelset_fracture(*frac_props[i], x));
+
+	std::vector<double> enrichments(frac_props.size() + junction_props.size());
+	//fractures possibly with branches
+	for (std::size_t i=0; i<frac_props.size(); i++)
+	{
+		auto const* frac = frac_props[i];
+		double enrich = levelsets[i];
+		for (std::size_t j=0; j<frac->branches_slave.size(); j++)
+			enrich *= Heaviside(levelset_branch(*frac->branches_slave[j], x));
+		enrichments[i] = enrich;
+	}
+
+	// junctions
+	for (std::size_t i=0; i<junction_props.size(); i++)
+	{
+		auto const* junction = junction_props[i];
+		auto fid1 = fracID_to_local.at(junction->fracture_IDs[0]);
+		auto fid2 = fracID_to_local.at(junction->fracture_IDs[1]);
+		double enrich = levelsets[fid1]*levelsets[fid2];
+		enrichments[i + frac_props.size()] = enrich;
+	}
+
+	return enrichments;
+}
+
+
+std::vector<double> du_global_enrichments(
+	std::size_t this_frac_id,
+	std::vector<FractureProperty*> const& frac_props,
+	std::vector<JunctionProperty*> const& junction_props,
+	std::unordered_map<int,int> const& fracID_to_local,
+	Eigen::Vector3d const& x)
+{
+	auto this_frac_local_index = fracID_to_local.at(this_frac_id);
```

Comment:
Not sure if bound checked access is intended; if not use `operator[]`.

Reply:
done.

Reply:
update: i remember it now. i had to use `at()` because `fracID_to_local` is const. `operator[]` cannot be used because it is non-const.